Advancements API

Cybermaxke opened the pull request:
**API** | [Common](https://github.com/SpongePowered/SpongeCommon/pull/1645) | [Forge](https://github.com/SpongePowered/SpongeForge/pull/1913) | [Vanilla](https://github.com/SpongePowered/SpongeVanilla/pull/363)

This PR adds the advancements API, with additions extra to the features provided in vanilla Minecraft. Implementations will follow when 1.12 is released.

The `AdvancementTree` represents a advancement Tab, it can be build with a root advancement. `Advancement`s that are attached to the root advancement will be present in the tree. The layout of the tree can also be customized, which means that you can move the advancements around in the tab though a `AdvancementTreeEvent.GenerateLayout` event.

`AdvancementCondition`s are the conditions that should be met before an advancement is unlocked. I provided AND and OR operations that can be applied. There is not limit in the complexity like in the json files.

There is also `Trigger`s that can be applied to the `AdvancementCondition`, the details of vanilla triggers are not exposed in the API (the configuration). However you can create your own `Trigger`s which will be parsed in the advancement json files.
http://minecraft.gamepedia.com/Advancements#Triggers

Additionally there is a `ScoreAdvancementCriterion` which requires a specific goal value to be achieved before it gets done, this is also supported by triggers. The trigger will need to trigger the criteria multiple times until the goal is reached, this is also supported in the json files by adding a `trigger_times` property to the trigger.

Extras:
* Generic Events - Used for the registrations of advancements and trees.
* `ScoreAdvancementCriterion` - Allows you to provide a goal value and when the score reaches its goal the advancement will be unlocked. Also supported in json files.
* Complex AND and OR condition operations
* Custom triggers (configurable) and supported in json
* Customize the advancement tree layout.

Cybermaxke committed:
Added the advancements API

stephan-gh commented on src/main/java/org/spongepowered/api/advancement/criteria/AdvancementCriterion.java:
```diff
@@ -0,0 +1,202 @@
+     * @param criteria The other criteria
+     * @return The and operation
+     */
+    static AdvancementCriterion buildAnd(AdvancementCriterion... criteria) {
```
I don't think the `build` prefix is needed here

Cybermaxke replied:
Otherwise will it conflict with the non-static `and` method.

mattmess1221 replied:
You can make it `static AdvancementCriterion and(AdvancementCriterion adv1, AdvancementCriterion... therest)`

kashike commented:
This should target the `1.12` branch.

kashike commented on src/main/java/org/spongepowered/api/advancement/Styled.java:
```diff
@@ -0,0 +1,155 @@
+ * Some that has a style within a advancement tree. Either an
+ * {@link AdvancementTree} or an {@link Advancement}.
+ */
+public interface Styled extends CatalogType, TextRepresentable {
```
This name should be more specific.

Cybermaxke commented:
@kashike I don't think that this PR will be finished before the 1.12 branch is merged. Or is there going to be separate development between 1.11 and 1.12 for a while?

kashike commented:
@Cybermaxke I'd say it's better to target `1.12` - there are changes there for 1.12, not in bleeding, and this PR targets 1.12.

stephan-gh commented:
@Cybermaxke You can easily change the target branch again once we merge 1.12. Until it is merged it should target 1.12 so you get all the updated dependencies etc. (you will also not be able to implement it by targeting bleeding)

mattmess1221 commented on src/main/java/org/spongepowered/api/advancement/AdvancementType.java:
```diff
@@ -0,0 +1,35 @@
+ * Changes the frame around the {@link Advancement} icon and
+ * also the appearance in the notifications.
+ */
+public interface AdvancementType extends CatalogType {
```
@ CatalogedBy(AdvancementTypes.class)

mattmess1221 commented on src/main/java/org/spongepowered/api/advancement/AdvancementTypes.java:
```diff
@@ -0,0 +1,61 @@
+
+    /**
+     * Notification appearance:
+     *      Challenge Complete!
```
Add a method for this in `AdvancementType`

Something like `List<Text> format(Advancement adv);`

Cybermaxke replied:
Should it be a `List`? Or maybe just a `Text` object where the lines are separated with `\n`.

sibomots replied:
Do not use embedded special characters.    Collection<E> where Collection is List, Set, etc.. is the right approach. Please, no magic format characters in the data.  Use the Library.

mattmess1221 commented on src/main/java/org/spongepowered/api/advancement/criteria/trigger/TriggerType.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Represents a source that can trigger a {@link AdvancementCriterion}.
+ */
+public interface TriggerType extends CatalogType {
```
@ CatalogedBy(TriggerTypes.class)

liach commented on src/main/java/org/spongepowered/api/advancement/AdvancementProgress.java:
```diff
@@ -0,0 +1,91 @@
+     * @param criterion The score criterion
+     * @return The score criterion progress
+     */
+    default ScoreCriterionProgress tryGet(ScoreAdvancementCriterion criterion) {
```
`getUnchecked`?

Cybermaxke replied:
Maybe `require` similar to the data methods?
https://github.com/SpongePowered/SpongeAPI/blob/bleeding/src/main/java/org/spongepowered/api/data/value/ValueContainer.java
`getUnchecked` isn't used anywhere else as far I know,

stephan-gh commented:
Please rebase your PR on the latest bleeding branch.

kashike commented:
@Cybermaxke Are you working on rebasing this?

Cybermaxke commented:
I will try to get back on this in a week or so, then I will have more time. I will also start implementing at that time.

randombyte-developer commented:
What's the status on this?

Cybermaxke commented:
@randombyte-developer I am going to start with the implementation, I already updated the API. ;)

parlough commented:
@Cybermaxke The hero we all want but don't deserve.

Cybermaxke committed:
Moved operator criteria to separate interfaces. Remove dataview method from trigger, it's not accessible in this form.

ImMorpheus commented on src/main/java/org/spongepowered/api/advancement/criteria/trigger/TriggerTypes.java:
```diff
@@ -0,0 +1,90 @@
+
+    public static final TriggerType CONSUME_ITEM = DummyObjectProvider.createFor(TriggerType.class, "CONSUME_ITEM");
+
+    public static final TriggerType CURSE_ZOMBIE_VILLAGER = DummyObjectProvider.createFor(TriggerType.class, "CONSUME_ITEM");
```
"CONSUME_ITEM" again ?

sibomots replied:
Typically, we use labels like Dummy to mean something that does absolutely nothing.  But here you're using it to do something meaningful.  Why not maker ObjectProvider generic   FooObjectProvider seems to be a factory.  See also, Design Patterns, GoF.  etc..

ImMorpheus commented on src/main/java/org/spongepowered/api/advancement/criteria/trigger/TriggerTypes.java:
```diff
@@ -0,0 +1,90 @@
+
+    public static final TriggerType VILLAGER_TRADE = DummyObjectProvider.createFor(TriggerType.class, "VILLAGER_TRADE");
+
+    public static final TriggerType CHANGED_DIMENSION = DummyObjectProvider.createFor(TriggerType.class, "CHANGED_DIMENSION");
```
This should be before CONSTRUCT_BEACON.

sibomots replied:
All of these string labels "VILLAGER_TRADE" should be moved into a class holding static final strings.  I am loathe to see embedded literal strings in code as much as possible.

liach replied:
just run sortfield once more and it will be fine.

ryantheleach replied:
@sibomots The strings here are used only for debugging. the dummy objects are nearly useless and as far as I'm aware don't even necessarily need correct names outside of logging in the case that something goes terribly wrong.

If Java had a way of copying the variable name or line number into a parameter that'd be the ideal solution.

Yes the dummy objects are a gross hack, but it's a gross hack that was decided on by the core team a long while ago, and isn't the subject of this PR outside correcting the incorrect debug text.

ImMorpheus commented:
Some triggers are missing, is that on purpose ? (effects_changed, levitation, ecc.)

Cybermaxke committed:
Changes for compatibility with vanilla mc. Simplify for impl.

Cybermaxke commented:
@ImMorpheus No, that's not intentional, they probably added more triggers after I designed the API.

sibomots commented:
AND, OR.. but not NOT ?

It's also remarkable how many bugs you might find and defects uncover by elaborating the JavaDocs.  Please attempt to flesh those out and scratch off the work item "java docs can be improved."  

Ie., writing the docs will uncover some defects and deliver aid for those who want to understand the PR.

Cybermaxke committed:
More work.

Cybermaxke commented:
@sibomots AND and OR operations are the only ones supported in vanilla minecraft. I already added a custom one for a "score" criterion. Do you know a usecase where a NOT operation could be used, this would complicate the internals a lot more so I will only added it if it's worth adding.

I will also document the javadocs a bit more when the implementation is finished, since things can still change.

liach commented on src/main/java/org/spongepowered/api/advancement/Advancer.java:
```diff
@@ -49,5 +49,5 @@
      * @return The advancement trees
      */
-    Set<AdvancementTree> getAdvancementTrees();
+    Set<AdvancementTree> getUnlockedAdvancementTrees();
```
Guess a collection may be more efficient

liach commented on src/main/java/org/spongepowered/api/advancement/criteria/AndCriterion.java:
```diff
@@ -43,4 +43,18 @@
+     * @param criteria The other criteria
+     * @return The AND operation
+     */
+    static AdvancementCriterion build(Iterable<AdvancementCriterion> criteria) {
```
Should you name them `of` to improve efficiency

liach replied:
I mean for shorter method names

liach commented on src/main/java/org/spongepowered/api/advancement/Advancement.java:
```diff
@@ -0,0 +1,133 @@
+     *
+     * @return The children advancements
+     */
+    List<Advancement> getChildren();
```
Do they have a particular order? If yes, can you specify in the javadoc? Otherwise, we may use a collection.

liach commented on src/main/java/org/spongepowered/api/advancement/Progressable.java:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @return Is achieved
+     */
+    default boolean achieved() {
```
You should leave this one. Implementations may check so that the method can return without creating extra objects.

liach commented on src/main/java/org/spongepowered/api/advancement/TreeLayoutElement.java:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @param position The position
+     */
+    void setPosition(Vector2d position);
```
I find it okay to redirect this one to `#setPosition(double, double)` like those methods in `Chunk` and `Entity`.

liach commented on src/main/java/org/spongepowered/api/advancement/criteria/AdvancementCriterion.java:
```diff
@@ -0,0 +1,145 @@
+     * Represents an empty criterion, this means that nothing has to be achieved to
+     * unlock a {@link Advancement}.
+     */
+    AdvancementCriterion EMPTY = DummyObjectProvider.createFor(AdvancementCriterion.class, "EMPTY");
```
I don't know if interface constants can be treated with access transformers. Can anyone enlighten me?

Cybermaxke replied:
It is done by setting a private static field, not through a access transformer (which is impossible for interfaces).
https://github.com/SpongePowered/SpongeCommon/blob/bleeding/src/main/java/org/spongepowered/common/registry/RegistryHelper.java#L96

liach commented on src/main/java/org/spongepowered/api/advancement/criteria/AdvancementCriterion.java:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @return The triggers
+     */
+    Set<Trigger> getTriggers();
```
I don't think we should use a set here. We don't need fast `#contains` operation, and we can indicate that there is no duplication in the javadoc. Triggers instances might not be kept very often; they only have their types exposed.

liach commented on src/main/java/org/spongepowered/api/advancement/criteria/OperatorCriterion.java:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @return The criteria
+     */
+    Set<AdvancementCriterion> getCriteria();
```
Why sets? I would suggest using naviagable sets and sort the criteria by their name to enable fast lookup and iteration (both)

liach commented on src/main/java/org/spongepowered/api/advancement/criteria/ScoreCriterionProgress.java:
```diff
@@ -0,0 +1,82 @@
+     * @param score The score to set
+     * @return The achieving time, if achieved
+     */
+    Optional<Instant> set(int score);
```
Does it mean setting the current score, and return the instant is the set score is the goal?

Cybermaxke replied:
If the goal was achieved before, that time will be returned, if you set a score lower than the goal, empty is returned, and equal to goal and not set before will return the current time.

parlough commented on src/main/java/org/spongepowered/api/event/advancement/CriterionRevokeEvent.java:
```diff
@@ -0,0 +1,31 @@
+
+import org.spongepowered.api.event.Cancellable;
+
+public interface CriterionRevokeEvent extends CriterionEvent, Cancellable {
```
I feel like elsewhere in the API we keep these as nested sub-terfaces to the original since they're so similar and it seems generally the `CriterionEvent` wouldn't be used on its own. Is there a particular reason we shouldn't here? 

For example look at `TabCompleteEvent` or `RconConnectionEvent` and its nested sub-interfaces. 

So if we make this change you'd listen to `CriterionEvent.Revoke` for this one. 

This would also apply to the advancement events.

Cybermaxke committed:
Updated the events, some improvements

- Don't use a specific collection type in the api
- Moved events to inner classes
- Replaced the advancement tree layout transformer with a event, this is the only moment that you can modify the layout.
- Minor changes.

Cybermaxke committed:
Fixed the SpongeEventFactoryTest for the added events.

Cybermaxke committed:
Progress. Added a game registry event.

Cybermaxke committed:
Added generic events.

gabizou commented on src/main/java/org/spongepowered/api/GameRegistry.java:
```diff
@@ -222,7 +223,10 @@
      * @throws CatalogTypeAlreadyRegisteredException if the type cannot be
      *      registered because a matching type was already registered
+     * @deprecated Is scheduled to be removed in API 8, the
+     *      {@link GameRegistryEvent.Register} should be used instead
```
Nested classes will not link properly in Java 8's javadoc generator. They need to be fully qualified to work.

Cybermaxke committed:
Fixed the SpongeEventFactoryTest.

Cybermaxke committed:
Make a Advancement extend TextRepresentable, moved the toast text method.

Cybermaxke committed:
Added a getTextFormat method to the advancement type.

tedstardev commented:
Do you think it would be possible to add a way to register and call custom triggers (likely namespaced by plugin ID) and criteria sets with this API?  I've simulated this in a vanilla MCP server but I'm not sure if it would be suitable/possible for Sponge?

For example, say I add a new trigger with a plugin: `banking:money_changed`
Then the server owner could add advancements that have the same trigger and a criteria (defined by the plugin) of say `"min": "500"` and it awards whatever advancement... maybe it would be useful?

Edit: Unless I'm totally missing something and you've already done that, I'm not too familiar with github or sponge api yet.. :/

Cybermaxke commented:
@tedstardev You can create custom criteria (with id) in the API and you can check for available trigger types. There is currently no API for custom triggers, but I will look into adding it.

The following test plugin demonstrates how you can use the API:
https://github.com/Cybermaxke/SpongeCommon/blob/e055c6e6c6f3c3b0b7e768e6a25bf1bfbbcd7b82/testplugins/src/main/java/org/spongepowered/test/AdvancementTest.java

Cybermaxke committed:
Work on custom triggers.

Cybermaxke committed:
Derp

XakepSDK commented:
How does api works?
Action item depends on advancement icon? Is it possible to set **stone** icon to advancement "smelt **iron**"
What is first arg of `build` method in test plugin?(in api i see only one string arg)

Also, make DUMMY criterion default.

Cybermaxke commented:
@XakepSDK You cannot modify other advancements with the API, if you really want to do that modify the vanilla resource pack, owever you can create new ones. And look at the javadocs if you want to know what each argument means, in the `build` methods case is it the identifier, without the namespace, this will be retrieved from the plugin container that is handling the event. Which means that you cannot register advancements with a different namespace.

XakepSDK commented:
Ah, I mean make my own advancement "smelt iron" with stone icon.

Cybermaxke commented:
@XakepSDK Ah yes, that is possible. The example should give a good idea how to do it.

Cybermaxke commented:
https://github.com/Cybermaxke/SpongeCommon/blob/e055c6e6c6f3c3b0b7e768e6a25bf1bfbbcd7b82/testplugins/src/main/java/org/spongepowered/test/AdvancementTest.java

XakepSDK commented:
For example - dig dirt. Where do you say, that player should "dig" dirt?
Only icon differs. Does advancement icon means action? E.g. shovel = dig,
furnace = smelt?

Cybermaxke commented:
@XakepSDK You trigger the advancements by listening to events for specific actions, for example within the break block event, join event, etc. You have to trigger them by manually. The icon is just visual, it has nothing to do with the triggering

Cybermaxke commented on src/main/java/org/spongepowered/api/advancement/AdvancementTree.java:
```diff
@@ -0,0 +1,98 @@
+         * @param id The identifier
+         * @return The advancement tree
+         */
+        AdvancementTree build(String id);
```
@gabizou Should I move to `id` to a separate method, to be consistent with the keys builder you added?

Cybermaxke committed:
Only one trigger per criterion.

Cybermaxke committed:
Moved the arguments out of the build methods to be consistent in the API.

Be consistent with added key builder.
